src/merkle_tree.rs

```diff
@@ -30,142 +47,131 @@ impl MerkleNode {
     }
 }
 
-impl<H: Hash + Clone> MerkleTree<H> {
+impl LeafNode {
+    pub fn new(hash_value: String) -> Self {
+        Self { hash_value }
+    }
+}
+
+impl<H: AsRef<[u8]> + Clone> MerkleTree<H> {
     pub fn new(transactions: Vec<H>) -> Result<Self, &'static str> {
         Self::create_tree(transactions)
     }
 
     // Fathers must have at least one son, if it does not have one, we clone the left one
     fn create_parent_from_siblings(
-        left_son: Box<MerkleNode>,
-        mut right_son: Option<Box<MerkleNode>>,
+        left_son: MerkleNode,
+        right_son: Option<MerkleNode>,
     ) -> MerkleNode {
-        let mut hasher = DefaultHasher::new();
-        left_son.hash_value.hash(&mut hasher);
-
-        match &right_son {
-            Some(right_sibling) => {
-                right_sibling.hash_value.hash(&mut hasher);
-            }
-            None => {
-                right_son = Some(left_son.clone());
-                left_son.hash_value.hash(&mut hasher);
-            }
-        }
-
-        MerkleNode::new(hasher.finish(), Some(left_son), right_son)
+        let mut hasher = Sha3_256::new();
```

Comment:
Maybe you can use the crate [digest](https://crates.io/crates/digest) instead and implement this for any type that implements it.
Then you could use `sha3_256` in tests to validate.